File: workers/go/harness/project.go
package harness

import (
	"context"
	"fmt"
	"net"
	"os"
	"os/signal"
	"syscall"

	"github.com/spf13/pflag"
	"github.com/temporalio/omes/workers/go/harness/api"
	sdkclient "go.temporal.io/sdk/client"
	"go.uber.org/zap"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

type ProjectRunMetadata struct {
	RunID       string
	ExecutionID string
}

type ProjectInitContext struct {
	Logger     *zap.SugaredLogger
	Run        ProjectRunMetadata
	TaskQueue  string
	ConfigJSON []byte
}

type ProjectExecuteContext struct {
	Logger    *zap.SugaredLogger
	Run       ProjectRunMetadata
	TaskQueue string
	Iteration int64
	Payload   []byte
}

type ProjectExecuteHandler func(sdkclient.Client, ProjectExecuteContext) error
type ProjectInitHandler func(sdkclient.Client, ProjectInitContext) error

type ProjectHandlers struct {
	Execute ProjectExecuteHandler
	Init    ProjectInitHandler
}

type projectServiceServer struct {
	api.UnimplementedProjectServiceServer

	handlers      ProjectHandlers
	clientFactory ClientFactory
	logger        *zap.SugaredLogger

	client sdkclient.Client
	run    *ProjectRunMetadata
}

func newProjectServiceServer(handlers ProjectHandlers, clientFactory ClientFactory, logger *zap.SugaredLogger) *projectServiceServer {
	if logger == nil {
		nop := zap.NewNop()
		logger = nop.Sugar()
	}
	return &projectServiceServer{
		handlers:      handlers,
		clientFactory: clientFactory,
		logger:        logger,
	}
}

func (s *projectServiceServer) Init(ctx context.Context, request *api.InitRequest) (*api.InitResponse, error) {
	if request.GetTaskQueue() == "" {
		return nil, status.Error(codes.InvalidArgument, "task_queue required")
	}
	if request.GetExecutionId() == "" {
		return nil, status.Error(codes.InvalidArgument, "execution_id required")
	}
	if request.GetRunId() == "" {
		return nil, status.Error(codes.InvalidArgument, "run_id required")
	}
	connectOptions := request.GetConnectOptions()
	if connectOptions == nil || connectOptions.GetServerAddress() == "" {
		return nil, status.Error(codes.InvalidArgument, "server_address required")
	}
	if connectOptions.GetNamespace() == "" {
		return nil, status.Error(codes.InvalidArgument, "namespace required")
	}

	config, err := buildClientConfig(clientConfigOptions{
		Logger:                  s.logger,
		ServerAddress:           connectOptions.GetServerAddress(),
		Namespace:               connectOptions.GetNamespace(),
		AuthHeader:              connectOptions.GetAuthHeader(),
		EnableTLS:               connectOptions.GetEnableTls(),
		TLSCertPath:             connectOptions.GetTlsCertPath(),
		TLSKeyPath:              connectOptions.GetTlsKeyPath(),
		TLSServerName:           connectOptions.GetTlsServerName(),
		DisableHostVerification: connectOptions.GetDisableHostVerification(),
	})
	if err != nil {
		return nil, status.Error(codes.InvalidArgument, err.Error())
	}

	client, err := s.clientFactory(config)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to create client: %v", err)
	}

	run := ProjectRunMetadata{
		RunID:       request.GetRunId(),
		ExecutionID: request.GetExecutionId(),
	}
	if s.handlers.Init != nil {
		if err := s.handlers.Init(client, ProjectInitContext{
			Logger:     s.logger,
			Run:        run,
			TaskQueue:  request.GetTaskQueue(),
			ConfigJSON: request.GetConfigJson(),
		}); err != nil {
			if client != nil {
				client.Close()
			}
			return nil, status.Errorf(codes.Internal, "init handler failed: %v", err)
		}
	}

	s.client = client
	s.run = &run
	return &api.InitResponse{}, nil
}

func (s *projectServiceServer) Execute(ctx context.Context, request *api.ExecuteRequest) (*api.ExecuteResponse, error) {
	if request.GetTaskQueue() == "" {
		return nil, status.Error(codes.InvalidArgument, "task_queue required")
	}
	if s.client == nil || s.run == nil {
		return nil, status.Error(codes.FailedPrecondition, "Init must be called before Execute")
	}
	if err := s.handlers.Execute(s.client, ProjectExecuteContext{
		Logger:    s.logger,
		Run:       *s.run,
		TaskQueue: request.GetTaskQueue(),
		Iteration: request.GetIteration(),
		Payload:   request.GetPayload(),
	}); err != nil {
		return nil, status.Errorf(codes.Internal, "execute handler failed: %v", err)
	}
	return &api.ExecuteResponse{}, nil
}

func runProjectServerCLI(handlers ProjectHandlers, clientFactory ClientFactory, argv []string) error {
	if handlers.Execute == nil {
		return fmt.Errorf("project handlers must provide an Execute handler")
	}
	if clientFactory == nil {
		return fmt.Errorf("client factory is required")
	}
	flagSet := pflag.NewFlagSet("project-server", pflag.ContinueOnError)
	port := flagSet.Int("port", 8080, "gRPC listen port")
	if err := flagSet.Parse(argv); err != nil {
		return err
	}
	logger, err := configureLogger("info", "console")
	if err != nil {
		return err
	}
	defer logger.Sync()

	listener, err := net.Listen("tcp", fmt.Sprintf("0.0.0.0:%d", *port))
	if err != nil {
		return err
	}
	server := grpc.NewServer()
	api.RegisterProjectServiceServer(server, newProjectServiceServer(handlers, clientFactory, logger))
	logger.Infof("Project server listening on port %d", *port)

	runContext, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer stop()

	serveErrCh := make(chan error, 1)
	go func() {
		serveErrCh <- server.Serve(listener)
	}()

	select {
	case err := <-serveErrCh:
		return err
	case <-runContext.Done():
		logger.Info("Shutdown signal received, stopping project server")
		server.GracefulStop()
		return <-serveErrCh
	}
}

File: workers/typescript/harness/src/project.ts
import { promisify } from 'node:util';
import * as grpc from '@grpc/grpc-js';
import { Client } from '@temporalio/client';
import { DefaultLogger, Logger } from '@temporalio/worker';
import { buildClientConfig, ClientFactory } from './client';
import type { ExecuteRequest__Output } from './generated/temporal/omes/projects/v1/ExecuteRequest';
import type { ExecuteResponse } from './generated/temporal/omes/projects/v1/ExecuteResponse';
import type { InitRequest__Output } from './generated/temporal/omes/projects/v1/InitRequest';
import type { InitResponse } from './generated/temporal/omes/projects/v1/InitResponse';
import { grpcError, registerProjectService } from './helpers';

export interface ProjectRunMetadata {
  runId: string;
  executionId: string;
}

export interface ProjectInitContext {
  logger: Logger;
  run: ProjectRunMetadata;
  taskQueue: string;
  configJson: Uint8Array;
}

export interface ProjectExecuteContext {
  logger: Logger;
  run: ProjectRunMetadata;
  taskQueue: string;
  iteration: number;
  payload: Uint8Array;
}

export type ProjectExecuteHandler = (
  client: Client,
  context: ProjectExecuteContext,
) => Promise<void>;

export type ProjectInitHandler = (client: Client, context: ProjectInitContext) => Promise<void>;

export interface ProjectHandlers {
  execute: ProjectExecuteHandler;
  init?: ProjectInitHandler;
}

const logger = new DefaultLogger('INFO');

export class ProjectServiceServer {
  private client: Client | undefined;
  private run: ProjectRunMetadata | undefined;

  constructor(
    private readonly handlers: ProjectHandlers,
    private readonly clientFactory: ClientFactory,
  ) {}

  async Init(request: InitRequest__Output): Promise<InitResponse> {
    const taskQueue = request.taskQueue ?? '';
    const executionId = request.executionId ?? '';
    const runId = request.runId ?? '';
    const connectOptions = request.connectOptions;

    if (!request.taskQueue) {
      throw grpcError(grpc.status.INVALID_ARGUMENT, 'task_queue required');
    }
    if (!executionId) {
      throw grpcError(grpc.status.INVALID_ARGUMENT, 'execution_id required');
    }
    if (!runId) {
      throw grpcError(grpc.status.INVALID_ARGUMENT, 'run_id required');
    }
    if (connectOptions === undefined || !connectOptions.serverAddress) {
      throw grpcError(grpc.status.INVALID_ARGUMENT, 'server_address required');
    }
    if (!connectOptions.namespace) {
      throw grpcError(grpc.status.INVALID_ARGUMENT, 'namespace required');
    }

    let config;
    try {
      config = buildClientConfig({
        server_address: connectOptions.serverAddress,
        namespace: connectOptions.namespace,
        auth_header: connectOptions.authHeader ?? '',
        tls: connectOptions.enableTls ?? false,
        tls_cert_path: connectOptions.tlsCertPath ?? '',
        tls_key_path: connectOptions.tlsKeyPath ?? '',
        tls_server_name: connectOptions.tlsServerName || undefined,
        disable_host_verification: connectOptions.disableHostVerification,
      });
    } catch (error) {
      throw grpcError(
        grpc.status.INVALID_ARGUMENT,
        error instanceof Error ? error.message : String(error),
      );
    }

    let client;
    try {
      client = await this.clientFactory(config);
    } catch (error) {
      throw grpcError(
        grpc.status.INTERNAL,
        `failed to create client: ${error instanceof Error ? error.message : String(error)}`,
      );
    }

    const run = {
      runId,
      executionId,
    };

    if (this.handlers.init !== undefined) {
      try {
        await this.handlers.init(client, {
          logger,
          run,
          taskQueue,
          configJson: request.configJson ?? Buffer.alloc(0),
        });
      } catch (error) {
        throw grpcError(
          grpc.status.INTERNAL,
          `init handler failed: ${error instanceof Error ? error.message : String(error)}`,
        );
      }
    }

    this.client = client;
    this.run = run;
    return {};
  }

  async Execute(request: ExecuteRequest__Output): Promise<ExecuteResponse> {
    const taskQueue = request.taskQueue ?? '';

    if (!taskQueue) {
      throw grpcError(grpc.status.INVALID_ARGUMENT, 'task_queue required');
    }
    if (this.client === undefined || this.run === undefined) {
      throw grpcError(grpc.status.FAILED_PRECONDITION, 'Init must be called before Execute');
    }

    try {
      await this.handlers.execute(this.client, {
        logger,
        run: this.run,
        taskQueue,
        iteration: request.iteration ?? 0,
        payload: request.payload ?? Buffer.alloc(0),
      });
    } catch (error) {
      throw grpcError(
        grpc.status.INTERNAL,
        `execute handler failed: ${error instanceof Error ? error.message : String(error)}`,
      );
    }

    return {};
  }
}

export async function runProjectServerCli(
  handlers: ProjectHandlers,
  clientFactory: ClientFactory,
  argv: readonly string[],
): Promise<void> {
  const port = parsePortArg(argv);
  const server = new grpc.Server();
  const bindAsync = promisify(server.bindAsync.bind(server));
  const tryShutdown = promisify(server.tryShutdown.bind(server));
  const service = new ProjectServiceServer(handlers, clientFactory);
  registerProjectService(server, service);
  await bindAsync(`0.0.0.0:${port}`, grpc.ServerCredentials.createInsecure());
  logger.info(`Project server listening on port ${port}`);

  await new Promise<void>((resolve, reject) => {
    process.once('SIGINT', () => {
      void tryShutdown().then(resolve, reject);
    });
  });
}

function parsePortArg(argv: readonly string[]): number {
  let port = 8080;
  for (let index = 0; index < argv.length; index += 1) {
    if (argv[index] === '--port') {
      port = Number.parseInt(argv[index + 1] ?? '', 10);
    }
  }
  return port;
}
